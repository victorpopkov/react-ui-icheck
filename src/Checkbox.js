import React, { Component } from 'react';
import PropTypes from 'prop-types';
import mobile from 'is-mobile';
import Input from './components/Input'; // eslint-disable-line sort-imports
import Label from './components/Label';
import deprecated from './helpers/deprecated';

class Checkbox extends Component {
  static propTypes = {
    activeClass: deprecated(PropTypes.string, 'activeClassName'),
    activeClassName: PropTypes.string,
    aria: PropTypes.bool,
    checkboxClass: deprecated(PropTypes.string, 'checkboxClassName'),
    checkboxClassName: PropTypes.string,
    checked: PropTypes.bool,
    checkedClass: deprecated(PropTypes.string, 'checkedClassName'),
    checkedClassName: PropTypes.string,
    children: PropTypes.node,
    className: PropTypes.string,
    defaultChecked: PropTypes.bool,
    determinateClass: PropTypes.string,
    disabled: PropTypes.bool,
    disabledClass: deprecated(PropTypes.string, 'disabledClassName'),
    disabledClassName: PropTypes.string,
    focusClass: deprecated(PropTypes.string, 'focusClassName'),
    focusClassName: PropTypes.string,
    hoverClass: deprecated(PropTypes.string, 'hoverClassName'),
    hoverClassName: PropTypes.string,
    id: PropTypes.string,
    increaseArea: PropTypes.string,
    indeterminate: PropTypes.bool,
    indeterminateClass: PropTypes.string,
    inheritClass: PropTypes.bool,
    inheritID: PropTypes.bool,
    insert: PropTypes.node,
    label: PropTypes.node,
    labelClassName: PropTypes.string,
    labelHover: PropTypes.bool,
    labelHoverClass: PropTypes.string,
    labelTag: PropTypes.string,
    labelTagClassName: PropTypes.string,
    name: PropTypes.string,
    type: PropTypes.string,
    value: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onMouseDown: PropTypes.func,
    onMouseLeave: PropTypes.func,
    onMouseUp: PropTypes.func,
    onTouchEnd: PropTypes.func,
    onTouchStart: PropTypes.func,
  };

  static defaultProps = {
    activeClass: 'active',
    activeClassName: 'active',
    aria: false,
    checkboxClass: 'icheckbox',
    checkboxClassName: 'icheckbox',
    checked: false,
    checkedClass: 'checked',
    checkedClassName: 'checked',
    children: null,
    className: null,
    defaultChecked: false,
    determinateClass: null,
    disabled: false,
    disabledClass: 'disabled',
    disabledClassName: 'disabled',
    focusClass: 'focus',
    focusClassName: 'focus',
    hoverClass: 'hover',
    hoverClassName: 'hover',
    id: null,
    increaseArea: '',
    indeterminate: true,
    indeterminateClass: 'indeterminate',
    inheritClass: false,
    inheritID: false,
    insert: null,
    label: null,
    labelClassName: null,
    labelHover: true,
    labelHoverClass: 'hover',
    labelTag: 'span',
    labelTagClassName: null,
    name: null,
    type: null,
    value: '',
    onBlur: null,
    onChange: null,
    onFocus: null,
    onMouseDown: null,
    onMouseLeave: null,
    onMouseUp: null,
    onTouchEnd: null,
    onTouchStart: null,
  };

  constructor(props) {
    super(props);

    this.handleHover = this.handleHover.bind(this);

    const { checked } = props;

    this.state = {
      checked: checked || false,
      hovered: false,
      isMobile: mobile(),
    };
  }

  componentWillReceiveProps(nextProps) {
    const { checked } = this.props;
    const { checked: nextChecked } = nextProps;

    if ('checked' in nextProps && checked !== nextChecked) {
      this.setState({
        checked: nextChecked,
      });
    }
  }

  handleHover(event, isHovered) {
    this.setState({
      hovered: isHovered,
    });
  }

  renderInput() {
    const {
      activeClass,
      activeClassName,
      aria,
      checkboxClass,
      checkboxClassName,
      checkedClass,
      checkedClassName,
      className,
      defaultChecked,
      disabled,
      disabledClass,
      disabledClassName,
      focusClass,
      focusClassName,
      hoverClass,
      hoverClassName,
      id,
      increaseArea,
      indeterminate,
      indeterminateClass,
      inheritClass,
      inheritID,
      insert,
      name,
      onBlur,
      onChange,
      onFocus,
      value,

      checked,
      children,
      determinateClass,
      label,
      labelClassName,
      labelHover,
      labelHoverClass,
      labelTag,
      labelTagClassName,
      onMouseDown,
      onMouseLeave,
      onMouseUp,
      onTouchEnd,
      onTouchStart,
      type,

      ...other
    } = this.props;
    const { checked: stateChecked, hovered, isMobile } = this.state;

    return (
      <Input
        {...other}
        activeClass={activeClass}
        activeClassName={activeClassName}
        aria={aria}
        checkboxClass={checkboxClass}
        checkboxClassName={checkboxClassName}
        checked={stateChecked}
        checkedClass={checkedClass}
        checkedClassName={checkedClassName}
        className={className}
        defaultChecked={defaultChecked}
        disabled={disabled}
        disabledClass={disabledClass}
        disabledClassName={disabledClassName}
        focusClass={focusClass}
        focusClassName={focusClassName}
        hoverClass={hoverClass}
        hoverClassName={hoverClassName}
        hovered={hovered}
        id={id}
        increaseArea={increaseArea}
        indeterminate={indeterminate}
        indeterminateClass={indeterminateClass}
        inheritClass={inheritClass}
        inheritID={inheritID}
        inputType="checkbox"
        insert={insert}
        isMobile={isMobile}
        name={name}
        ref={(c) => { this.input = c; }}
        value={value}
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
      />
    );
  }

  render() {
    const {
      children,
      disabled,
      label,
      labelClassName,
      labelHover,
      labelTag,
      labelTagClassName,
    } = this.props;
    const { isMobile } = this.state;

    if (!label && !children) {
      return this.renderInput();
    }

    return (
      <Label
        disabled={disabled}
        handleHover={this.handleHover}
        isMobile={isMobile}
        label={label}
        labelChildren={children}
        labelClassName={labelClassName}
        labelHover={labelHover}
        labelTag={labelTag}
        labelTagClassName={labelTagClassName}
      >
        {this.renderInput()}
      </Label>
    );
  }
}

export default Checkbox;
