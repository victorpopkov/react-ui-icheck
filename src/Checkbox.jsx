import PropTypes from 'prop-types';
import BaseInput from './BaseInput'; // eslint-disable-line sort-imports

class Checkbox extends BaseInput {
  renderInput() {
    const { checkboxClass, checkboxClassName } = this.props;

    return super.renderInput({
      checkboxClassName: checkboxClass || checkboxClassName,
      inputType: 'checkbox',
    });
  }
}

Checkbox.propTypes = {
  ...BaseInput.propTypes,
  checkboxClassName: PropTypes.string,
};

Checkbox.defaultProps = {
  ...BaseInput.defaultProps,
  checkboxClass: null,
  checkboxClassName: 'icheckbox',
};

export default Checkbox;
