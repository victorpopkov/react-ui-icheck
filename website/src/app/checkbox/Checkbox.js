import { Checkbox, CheckboxGroup } from 'react-ui-icheck'; // eslint-disable-line import/no-unresolved
import {
  Col,
  Container,
  Nav,
  NavItem,
  NavLink,
  Row,
  TabContent,
  TabPane,
} from 'reactstrap';
import React, { Component } from 'react';
import classNames from 'classnames';
import { hot } from 'react-hot-loader';
import { Example, Jumbotron, Sidebar } from '../common'; // eslint-disable-line sort-imports
import CheckboxProperties from './CheckboxProperties';

@hot(module)
class CheckboxPage extends Component {
  constructor(props) {
    super(props);

    this.toggleActiveTab = this.toggleActiveTab.bind(this);

    this.state = {
      activeTab: 'examples',
    };
  }

  toggleActiveTab(tab) {
    const { activeTab } = this.state;

    if (activeTab !== tab) {
      this.setState({ activeTab: tab });
    }
  }

  render() {
    const { activeTab } = this.state;

    return (
      <main role="main">
        <Jumbotron />
        <Container>
          <Row>
            <Col md={9}>
              <h2 className="heading">Checkbox</h2>
              <Nav className="mt-4" tabs>
                <NavItem>
                  <NavLink
                    className={classNames({ active: activeTab === 'examples' })}
                    onClick={() => { this.toggleActiveTab('examples'); }}
                  >
                    Examples
                  </NavLink>
                </NavItem>
                <NavItem>
                  <NavLink
                    className={classNames({ active: activeTab === 'properties' })}
                    onClick={() => { this.toggleActiveTab('properties'); }}
                  >
                    Properties
                  </NavLink>
                </NavItem>
              </Nav>
              <TabContent
                activeTab={activeTab}
                className="mb-4"
              >
                <TabPane tabId="examples">
                  <Example
                    srcPrepend={`import 'icheck/skins/all.css';
import { Checkbox, CheckboxGroup } from 'react-ui-icheck';\n\n`}
                    title="Default"
                  >
                    <CheckboxGroup
                      checkboxWrapClassName="form-check"
                      checkboxWrapTag="div"
                    >
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        increaseArea="30%"
                        label="#checkbox-1"
                      />
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        increaseArea="30%"
                        label={'#checkbox-2, <span class="label1">defaultChecked</span>'}
                        labelTag="div"
                        labelTagClassName="d-inline"
                        defaultChecked
                      />
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        increaseArea="30%"
                        labelTag="div"
                        labelTagClassName="d-inline"
                        disabled
                      >
                        <span>#checkbox-3, </span>
                        <span className="label1">disabled</span>
                      </Checkbox>
                    </CheckboxGroup>
                  </Example>
                  <Example
                    srcPrepend={`import 'icheck/skins/all.css';
import { Checkbox, CheckboxGroup } from 'react-ui-icheck';\n\n`}
                    title="Inline"
                  >
                    <CheckboxGroup
                      checkboxWrapClassName="form-check form-check-inline"
                      checkboxWrapTag="div"
                      className="d-flex align-items-center"
                    >
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        label="#checkbox-1"
                      />
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        label={'#checkbox-2, <span class="label1">defaultChecked</span>'}
                        labelTag="div"
                        labelTagClassName="d-inline"
                        defaultChecked
                      />
                      <Checkbox
                        checkboxClass="icheckbox_square-blue"
                        labelTag="div"
                        labelTagClassName="d-inline"
                        disabled
                      >
                        <span>#checkbox-3, </span>
                        <span className="label1">disabled</span>
                      </Checkbox>
                    </CheckboxGroup>
                  </Example>
                </TabPane>
                <TabPane tabId="properties">
                  <CheckboxProperties />
                </TabPane>
              </TabContent>
            </Col>
            <Sidebar />
          </Row>
        </Container>
      </main>
    );
  }
}

export default CheckboxPage;
