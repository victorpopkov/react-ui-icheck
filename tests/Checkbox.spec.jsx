import React from 'react';
import expect from 'expect';
import { fireEvent, render } from '@testing-library/react';
import Checkbox from '../src/Checkbox';

describe('<Checkbox />', () => {
  describe('when no props are passed', () => {
    it('should match the snapshot', () => {
      const { container } = render(<Checkbox />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('when `label` prop is "test"', () => {
    it('should have <span /> with "test" text', () => {
      const { container } = render(<Checkbox label="test" />);
      expect(container.querySelector('span').innerHTML).toBe('test');
    });

    describe('when hovering <label />', () => {
      it('<div class="icheckbox" /> should have "hover" class while being hovered', () => {
        const { container } = render(<Checkbox label="test" />);
        const icheckboxElement = container.querySelector('.icheckbox');
        const labelElement = container.querySelector('label');
        fireEvent.mouseEnter(labelElement);
        expect(icheckboxElement).toHaveClass('hover');
        fireEvent.mouseLeave(labelElement);
        expect(icheckboxElement).not.toHaveClass('hover');
      });
    });

    describe('when touching <label />', () => {
      it('<div class="icheckbox" /> should have "hover" class while being touched', () => {
        const { container } = render(<Checkbox label="test" />);
        const icheckboxElement = container.querySelector('.icheckbox');
        const labelElement = container.querySelector('label');
        fireEvent.touchStart(labelElement);
        expect(icheckboxElement).toHaveClass('hover');
        fireEvent.touchEnd(labelElement);
        expect(icheckboxElement).not.toHaveClass('hover');
      });
    });
  });

  describe('when `checked` prop is `true`', () => {
    it('<input /> should have "checked" attribute', () => {
      const { container } = render(<Checkbox checked />);
      expect(container.querySelector('input')).toHaveAttribute('checked');
    });

    it('<div class="icheckbox" /> should have "checked" class', () => {
      const { container } = render(<Checkbox checked />);
      expect(container.querySelector('.icheckbox')).toHaveClass('checked');
    });
  });

  describe('when focusing <input />', () => {
    it('<div class="icheckbox" /> should have "focus" class while being focused', () => {
      const { container } = render(<Checkbox />);
      const icheckboxElement = container.querySelector('.icheckbox');
      const inputElement = container.querySelector('input');
      fireEvent.focus(inputElement);
      expect(icheckboxElement).toHaveClass('focus');
      fireEvent.blur(inputElement);
      expect(icheckboxElement).not.toHaveClass('focus');
    });
  });

  describe('when children are passed', () => {
    it('should match the snapshot', () => {
      const { container } = render(<Checkbox>label</Checkbox>);
      expect(container).toMatchSnapshot();
    });

    describe('and `labelTag` prop is "div"', () => {
      it('should match the snapshot', () => {
        const { container } = render(
          <Checkbox labelTag="div">
            <span>label</span>
          </Checkbox>,
        );
        const divElements = container
          .querySelector('label')
          .querySelectorAll('div');
        expect(divElements[1]).toMatchSnapshot();
      });
    });
  });
});
