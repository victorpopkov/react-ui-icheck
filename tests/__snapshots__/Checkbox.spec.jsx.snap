// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component and the \`label\` prop is \`label\` as render should match the snapshot 1`] = `
<label>
  <div
    class="icheckbox indeterminate"
  >
    <input
      style="opacity:0;position:absolute"
      type="checkbox"
      value=""
    />
    <ins
      class="iCheck-helper"
      role="presentation"
      style="position:absolute;top:0%;left:0%;display:block;width:100%;height:100%;margin:0;padding:0;background:#fff;border:0;opacity:0;cursor:pointer"
    />
  </div>
  <span>
    label
  </span>
</label>
`;

exports[`Checkbox component and the children are passed and the \`labelTag\` prop is \`div\` as render should match the snapshot 1`] = `
<label>
  <div
    class="icheckbox indeterminate"
  >
    <input
      style="opacity:0;position:absolute"
      type="checkbox"
      value=""
    />
    <ins
      class="iCheck-helper"
      role="presentation"
      style="position:absolute;top:0%;left:0%;display:block;width:100%;height:100%;margin:0;padding:0;background:#fff;border:0;opacity:0;cursor:pointer"
    />
  </div>
  <div>
    <span>
      label
    </span>
  </div>
</label>
`;

exports[`Checkbox component as mount should match the default props snapshot 1`] = `
{
  "activeClassName": "active",
  "aria": false,
  "checkboxClass": null,
  "checkboxClassName": "icheckbox",
  "checked": false,
  "checkedClassName": "checked",
  "children": null,
  "className": null,
  "disabled": false,
  "disabledClassName": "disabled",
  "focusClassName": "focus",
  "hoverClassName": "hover",
  "id": null,
  "increaseArea": "",
  "indeterminate": true,
  "indeterminateClassName": "indeterminate",
  "inheritClassName": false,
  "inheritID": false,
  "insert": null,
  "label": null,
  "labelClassName": null,
  "labelHover": true,
  "labelHoverClass": "hover",
  "labelTag": "span",
  "labelTagClassName": null,
  "name": null,
  "onBlur": null,
  "onChange": null,
  "onFocus": null,
  "onMouseDown": null,
  "onMouseLeave": null,
  "onMouseUp": null,
  "onTouchEnd": null,
  "onTouchStart": null,
  "type": null,
  "value": "",
}
`;

exports[`Checkbox component as mount should match the default state snapshot 1`] = `
{
  "checked": false,
  "hovered": false,
  "isMobile": false,
  "prevPropChecked": false,
}
`;

exports[`Checkbox component as render should match the snapshot 1`] = `
<div
  class="icheckbox indeterminate"
>
  <input
    style="opacity:0;position:absolute"
    type="checkbox"
    value=""
  />
  <ins
    class="iCheck-helper"
    role="presentation"
    style="position:absolute;top:0%;left:0%;display:block;width:100%;height:100%;margin:0;padding:0;background:#fff;border:0;opacity:0;cursor:pointer"
  />
</div>
`;
